Use a summary table for PR comments

Kobzol opened the pull request:
This PR implements a new design for reporting performance summaries on PRs.

It looks something like this:
| | Regressions 😿 <br />(primary) | Regressions 😿 <br />(secondary) | Improvements 🎉 <br />(primary) | Improvements 🎉 <br />(secondary) | All 😿 🎉 <br />(primary) |
|:---:|:---:|:---:|:---:|:---:|:---:|
| count[^1] | 2 | 1 | 2 | 1 | 4 |
| mean[^2] | 150.0% | 100.0% | -62.5% | -66.7% | 43.8% |
| max | 200.0% | 100.0% | -75.0% | -66.7% | 200.0% |

[^1]: *number of relevant changes*
[^2]: *the arithmetic mean of the percent change*

Should the table also be used in the weekly triage or not? (will be left for another PR).

Closes: https://github.com/rust-lang/rustc-perf/issues/1214

rylev reviewed (changes requested):
Looks like you're missing a run of `cargo fmt` to get CI passing.

> 1. Should the table also be used in the weekly triage or not?

Yes I think that would be helpful

> 2. I don't really like that write_summary_table is a method on ComparisonSummary, this should be a free function I think. Same with write_summary_lines.

Feel free to make this refactor 😊

> 3. Currently, primary and secondary benchmarks are not distinguished, because this information is not currently propagated to the performance summaries. Should we add the benchmark configuration to TestResultComparison or pass some Map<BenchmarkName, BenchmarkData> to the summary?

I'd like for this change to get in before we merge if possible. It seems somewhat redundant to put the category in with the `TestResultComparison` since that data will be the same for all test cases with the same benchmark. So having some `BenchMark` to `Category` map probably would work best. 

> 4. Should columns without any entries be present or not? I.e. should we show 0 entries or just hide the column altogether?

I think we should keep all columns even when they have 0 entries.

Kobzol commented:
To avoid complicating the table render function and to keep category handling out of summaries, I changed it so that the render function directly receives two summaries (primary and secondary) and it is up to the caller to generate these. Since the summary can now be empty, I commented the check that returned `None` from `summarize_comparison` if the resulting summary would be empty.

One issue is that some columns are split between primary/secondary, but the last one (all relevant changes) contains data for both categories together. I added a TODO comment to the code with possible ways to resolve this, I'll be glad to hear your thoughts.

rylev commented on site/src/comparison.rs:
```diff
@@ -193,9 +194,9 @@ impl ComparisonSummary {
-        if comparisons.len() == 0 {
-            return None;
-        }
+        // if comparisons.len() == 0 {
+        //     return None;
+        // }
```
I think this needs to be reverted.

Kobzol replied:
Then the summary table would need to take `Option<ComparisonSummary>` instead of `ComparisonSummary`, which would make the code a bit more complicated. It seems better to me to handle the case "no data available" without any special case.

But we can also solve it by passing an empty summary to the table function if this function returned None, we just need to add a constructor that creates an empty summary.

rylev replied:
Ok that's fine. Then it just needs to be deleted (instead of commented out).

rylev commented on site/src/comparison.rs:
```diff
@@ -421,61 +422,103 @@ impl ComparisonSummary {
+
+    writeln!(
+        result,
+        r#"| | Regressions 😿 <br> (primary) | Regressions 😿 <br> (secondary) | Improvements 🎉 <br> (primary) | Improvements 🎉 <br> (secondary) | All relevant changes |
```
I think if we keep the "All relevant changes" column it should just be the primary benchmarks.

Kobzol replied:
Okay, I think I agree. But before I change it, @Mark-Simulacrum @nnethercote @lqd what are your thoughts on this?

lqd replied:
Having it summarize the primary benchmarks also seems sensible to me, to de-emphasize the secondary benchmark results (and their data is still there next to it anyways), but I myself mostly focus on the magnitude of regressions and improvements individually rather than the complete aggregate results.

rylev commented on site/src/comparison.rs:
```diff
@@ -1291,83 +1334,130 @@ fn compare_link(start: &ArtifactId, end: &ArtifactId) -> String {
-| count[^1] | 0 | 2 | 2 |
-| mean[^2] | N/A | -62.5% | -62.5% |
-| max | N/A | -75.0% | 0.0% |
+    /*#[test]
```
Don't forget to revert this

rylev reviewed (changes requested):
> the `Category` is not exported by collector (this was done intentionally as I remember)

Why doesn't the collector export `Category`?

> One issue is that some columns are split between primary/secondary

It's a shame that GitHub doesn't support more complex tables. If it did we could have the category be a two column wide column that splits into regressions and improvements. Oh well.

I think it's much more interesting to have the average change for the primary category. I think mixing primary and secondary will make for worse data.

Kobzol commented:
> Why doesn't the collector export `Category`?

I checked `git` history. It was originally in database:
https://github.com/rust-lang/rustc-perf/pull/1181/files#diff-9c399cba4cb055f29cca4a48993625e40ffe47dd9a434ef8de6024d72dcdd1b9R775

Then it was moved in https://github.com/rust-lang/rustc-perf/commit/4009f1e8bf3ef851f96654ab5b2fc6b8ff94e70f (I don't think that this had a PR). @nnethercote would it be okay if `Category` was exported from the `collector` crate?

nnethercote commented:
Exporting `Category` seems fine to me.

Kobzol commented:
Exported `Category` and hooked the summary table function to the GitHub listener.

Kobzol commented:
I would leave changing the summary used in triages for another PR, there are some links being generated there and I'm not sure how to make that work with the table. But for PR comments this should be ready for review.

rylev commented on site/src/comparison.rs:
```diff
@@ -311,6 +320,13 @@ impl ComparisonSummary {
         self.arithmetic_mean(self.comparisons.iter())
     }
 
+    pub fn significant_changes(&self) -> usize {
```
nit: can we call this `num_significant_changes`?

rylev commented on site/src/github.rs:
```diff
@@ -673,71 +691,106 @@ async fn categorize_benchmark(
+        a: comparison.a,
+        b: comparison.b,
+        statistics: secondary,
+        new_errors: comparison.new_errors.clone(),
```
Nit: the clone call isn't necessary

rylev commented on site/src/comparison.rs:
```diff
@@ -423,6 +439,102 @@ impl ComparisonSummary {
+
+    writeln!(
+        result,
+        r#"| | Regressions 😿 <br> (primary) | Regressions 😿 <br> (secondary) | Improvements 🎉 <br> (primary) | Improvements 🎉 <br> (secondary) | All relevant primary changes |
```
Instead of "All relevant primary changes" which renders very wide and makes the table as a whole very wide, perhaps we can do "All 😿🎉 \<br\> (primary)". That should make the table significantly less wide.

rylev commented on site/src/github.rs:
```diff
@@ -673,71 +691,106 @@ async fn categorize_benchmark(
+    let num_regressions = relevant_summary.number_of_regressions();
 
-    let description = match direction {
+    let short_summary = match direction {
```
I'm not sure how I feel about the short summary. We're dropping one of the categories (either primary or secondary) and not even mentioning which of the two we're summarizing.

Perhaps for now we can just summarize both:

Summary:
* Primary benchmarks: 🎉 relevant improvements found
* Secondary benchmarks: no relevant changes found

Kobzol replied:
Good idea! I'll change it like this.

rylev reviewed (changes requested):
I think we're almost there but there's a few things I'd like addressed.

Kobzol commented:
Now that I look at the [table](https://github.com/Kobzol/rustc-perf/issues/1#issuecomment-1079242751), maybe it would be better to group the columns by category and not by direction? `primary - imp`, `primary - reg`, `secondary - imp`, `secondary - reg`?

rylev commented:
I imagine that folks are generally more interested in regressions (i.e., improvements are great but for PRs that are not explicitly performance related most people just want to know "did this change make the compiler slower?"). With that in mind I think it makes most sense to show regressions first and then improvements (i.e., group by direction)

rylev commented on site/src/comparison.rs:
```diff
@@ -423,6 +439,102 @@ impl ComparisonSummary {
+    if !primary.errors_in.is_empty() {
+        write!(
+            result,
+            "- Benchmark(s) {} started failing to build",
```
I think it might be better do the following as I think it might be hard to read the way it's phrased now:

The following benchmarks failed to build:
- serde-1.0.135
- syn-1.0.89

Kobzol replied:
Good point, I changed it.

rylev reviewed (approved):
Looks good. I had one nit, but we can merge this first before addressing that if you want to

rylev reviewed (approved):
🎉

lqd commented:
@Kobzol the report looks strange https://github.com/rust-lang/rust/pull/95171#issuecomment-1085849415 but you must have noticed that's your PR :)

Kobzol commented:
Thanks, I'll check it ASAP.
